### ISSUE TITLE
Incorrect ARIA role/behavior of paper-tabs

### ISSUE THREAD

**Issue by Sunlis** [PolymerElements/paper-tabs#96]
There is an ARIA spec for tabs, which defined tab order and arrow key behavior: http://www.w3.org/TR/wai-aria-practices/#tabpanel
Paper-tabs (and paper-tab) is marked correctly for this, but doesn't match the behavior. Paper-tabs follows the behavior for the menu spec: http://www.w3.org/TR/wai-aria-practices/#menu

Paper-tabs should either: change roles to menu/menuitem; or respond to arrow keys by changing tabs rather than changing focus.

**Comment by bicknellr** [comment]
Thanks for pointing this out. I think we should activate tabs on moving to them rather than changing the role to menu.

**Comment by bicknellr** [comment]
@alice, what do you think about this?

**Comment by alice** [comment]
Let me ask around about expectations around tabpanels. The ARIA design patterns are really good but just occasionally don't match reality, so I like to double check.

**Comment by alice** [comment]
Ok, straw poll of 2 suggests that the design patterns are spot on in this case, so I agree with @bicknellr that this should be implemented :grinning:

**Comment by alice** [comment]
https://www.youtube.com/watch?v=Maa73HqDsSw shows this UI pattern in action in OS X settings.

**Comment by bicknellr** [comment]
@alice Is there any case where focusing a tab wouldn't be expected to activate it?

**Comment by alice** [comment]
Are all tabs focusable at all times? Or does it follow the ARIA pattern guidance of having only the active tab focusable?

**Comment by bicknellr** [comment]
The active tab is the only one with tabindex 0, all the others have tabindex -1; so they're all focusable, but tabbing wouldn't focus them.

**Comment by alice** [comment]
This seems reasonable to me. Clicking would also focus, but clicking would also activate the tab so
¯\_(ツ)_/¯

**Comment by bicknellr** [comment]
Cool, this should be ready once I figure out why Firefox isn't happy.

**Comment by robrez** [comment]
I could see this making data-bindings go a nuts if the events aren't debounced a bit.
eg: neon-animated-pages bound to the selected tab, user is on tab1 and wants tab3 of 6... several page switches are initiated

**Comment by bicknellr** [comment]
Yeah, that's a good point.

**Comment by alice** [comment]
Agreed. I wonder if this pattern needs to be modified slightly for the web.

**Comment by bicknellr** [comment]
After playing around with a few sets of tabs in OS X and Windows 10, I wouldn't even go so far as to call it an established 'pattern' lul - native UIs seem to be really inconsistent about this. Some switch immediately on focus, some debounce, some wait for space and/or enter, some aren't even focusable via the keyboard at all. IMHO, the OS X trackpad settings example you posted earlier (automatic switching after short debounce) feels nicest but waiting for space / enter also seems like an acceptable fallback.

**Comment by Sunlis** [comment]
Events shouldn't be debounced by the sender because there are listeners that do expensive things. The listeners doing the expensive stuff should be the ones doing the denouncing.

**Comment by bicknellr** [comment]
The debounce is meant more to prevent thrashing given that a user switching tabs via the keyboard might not be able to switch directly from one tab to the other without traversing all the tabs in between.

**Comment by robrez** [comment]
@Sunlis That's a fair point and I can't say I disagree, generally.

A big part of the Polymer's appeal is the data-binding system. Requiring listeners to do the debouncing could force developers to opt out of data-bindings in many cases. I've seen a handful of elements that utilize a configurable debounce-threshold property for one reason or another.. Here are a couple the come to mind:

element | property
----------|---------------
iron-ajax | debounceDuration
iron-page-url | dwellTime

Maybe something similar here could make everyone happy?

**Comment by bicknellr** [comment]
(Also, just to be clear, I don't think we should delay tab activation when clicking / tapping.)

**Comment by Sunlis** [comment]
Possibly my biggest gripe with Polymer is that a lot of elements make assumptions like "just debounce the events", which sounds great but screws the user that doesn't want debounced events.

A configurable threshold would be good here, with some reasonable default to prevent trashing. As long as I can disable the debounce by setting it to `0`. :wink:

**Comment by bicknellr** [comment]
Ok, configurable + ability to opt out (0) sounds good to me.
